relations/src/shielder/tangle.rs

```diff
@@ -123,15 +150,64 @@ fn _tangle(bytes: &mut [u8], low: usize, high: usize) {
         }
     } else {
         let mid = (low + high) / 2;
-        _tangle(bytes, low, mid);
-        _tangle(bytes, mid, high);
+        do_tangle(elems, low, mid);
+        do_tangle(elems, mid, high);
 
         for i in low..mid {
-            bytes.swap(i, i + mid - low);
+            elems.swap(i, i + mid - low);
         }
 
         for i in low + 1..high {
-            bytes[i] = u8::overflowing_mul(bytes[i], bytes[i - 1]).0;
+            let product = elems[i] * elems[i - 1];
+            elems[i] = dezeroize(&product, (low * high + i) as u64);
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use ark_ff::Zero;
+    use ark_r1cs_std::{fields::FieldVar, R1CSVar};
+
+    use crate::{
+        environment::FpVar,
+        shielder::tangle::{tangle, tangle_in_circuit},
+        CircuitField,
+    };
+
+    #[test]
+    fn tangling_is_homomorphic() {
+        let input = vec![
+            CircuitField::from(0u64),
+            CircuitField::from(100u64),
+            CircuitField::from(17u64),
+            CircuitField::from(19u64),
+        ];
```

Comment:
Out of curiosity - is there anything specific in this example that makes you confident that tangling is homomorphic?